Sort the output of ClusterTPAssociationProducer

makortel opened the pull request:
While preparing #9223, I noticed that the ClusterTPAssociationList is copied in QuickTrackAssociatorByHitsImplProducer (to sort it), and the inserted to Event as part of reco::TrackToTrackingParticleAssociator. IgProf shows that in RelVal TTbar+35BX@25ns sample this takes O(4-5 MB) of memory in a RAW2DIGI,RECO,VALIDATION,DQM workflow.

In this PR I try to avoid the copy by sorting ClusterTPAssociationList in ClusterTPAssociationProducer before the product is put in Event.

On the same go, I turned the shared_ptr<TrackerHitAssociator> to unique_ptr as the TrackerHitAssociator is not shared with other objects, and moved to C++14 make_unique.

Tested in CMSSW_7_5_X_2015-05-13-2300 and CMSSW_7_5_X_2015-05-21-2300, no changes expected in results.

@rovere @VinInn

makortel committed:
Move ClusterTPAssociationList sorting from QuickTrackAssociatorByHitsProducer to ClusterTPAssociationProducer

This avoids making a copy of ClusterTPAssociationList in
QuickTrackAssociatorByHitsProducer.

Turned one shared_ptr to unique_ptr (as the pointer is not shared with
other objects) and moved to make_unique on the same go.

cmsbuild commented:
A new Pull Request was created by @makortel (Matti Kortelainen) for CMSSW_7_5_X.

Sort the output of ClusterTPAssociationProducer

It involves the following packages:

SimTracker/TrackAssociatorProducers
SimTracker/TrackerHitAssociation


@cmsbuild, @civanch, @nclopezo, @mdhildreth can you please review it and eventually sign? Thanks.
@GiacomoSguazzoni, @rovere, @VinInn, @appeltel, @wmtford, @cerati, @threus, @dgulhan, @venturia this is something you requested to watch as well.
You can sign-off by replying to this message having '+1' in the first line of your reply.
You can reject by replying  to this message having '-1' in the first line of your reply.
If you are a L2 or a release manager you can ask for tests by saying 'please test' in the first line of a comment.
@nclopezo you are the release manager for this.
You can merge this pull request by typing 'merge' in the first line of your comment.

makortel commented:
@Dr15Jones @mark-grimes as the previous developers on QuickTrackAssociatorByHits+related code, could you please take a look if this ok?

civanch commented:
please test

cmsbuild commented:
The tests are being triggered in jenkins.

cmsbuild commented:
+1
Tested at: cc7abe4c1799c0953833ed0d85e65e28a373234c
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9225/4955/summary.html

Dr15Jones commented on SimTracker/TrackAssociatorProducers/plugins/QuickTrackAssociatorByHitsImpl.h:
```diff
@@ -5,6 +5,8 @@
 
 #include "SimDataFormats/Associations/interface/TrackToTrackingParticleAssociator.h"
 
+#include "SimTracker/TrackerHitAssociation/interface/ClusterTPAssociationProducer.h"
```
I believe it is bad policy to ever allow other code to read the header file for a producer. I strongly suggest moving the typedef somewhere else.

Dr15Jones replied:
OK, it isn't as bad here since it is still contained within the plugins directory, but on general principles I think what I said is still true, nothing should ever depend on a producer's code.

makortel committed:
Move ClusterTPAssociationList typedef to a separate header

In order to avoid #including an EDProducer header in QuickTrackAssociatorByHitsImpl/Producer

makortel commented:
Thanks @Dr15Jones, I moved the definition of ClusterTPAssociationList to a separate header (still in SimTracker/TrackerHitAssociation as the corresponding dictionaries are defined in there; is that actually a proper place for them, i.e. outside of any *DataFormats?).

cmsbuild commented:
Pull request #9225 was updated. @cmsbuild, @civanch, @nclopezo, @mdhildreth can you please check and sign again.

cmsbuild commented:
+1
Tested at: ce21d65ec5a5c554d253bca30d529c762bbaa911
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9225/50/summary.html

cmsbuild commented:
Comparison is ready
https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-9225/50/summary.html

civanch commented:
+1

cmsbuild commented:
This pull request is fully signed and it will be integrated in one of the next CMSSW_7_5_X IBs unless changes (tests are also fine). This pull request requires discussion in the ORP meeting before it's merged. @davidlange6, @nclopezo, @smuzaffar

mark-grimes commented:
@makortel, I had a look through this PR.  To be honest, the layout has changed so much since I worked on it that I barely recognise it anymore.  As far as I can tell from my limited knowledge it looks fine.

makortel commented:
Thanks @mark-grimes. So to your knowledge the idea of doing the sorting in ClusterTPAssociationProducer should not have any non-trivial consequences (that are not visible in the tests done so far)?

Dr15Jones commented:
@makortel Given that ```ClusterTPAssociationList ``` is an internal detail of the algorithm, I personally would have put the header in SimTracker/TrackAssociatorProducers/plugins and referred to it without any preceeding path.

mark-grimes commented:
Not as far as I can see, but I don't think I'm qualified to say for sure.

davidlange6 commented:
+1

makortel commented:
@Dr15Jones The typedef and the sorting function is used also in `SimTracker/TrackerHitAssociation/plugins`. My feeling is that including a header from other package's `plugins` would be even worse than what is done here. Of course if I've understood incorrectly, I'll be happy to move the header (in another PR).

Dr15Jones commented:
I had missed the fact that the header was being used in two different packages. I also agree that one should never include a header coming from a plugins directory.
